NetscapeBookmarkParser.php

```diff
@@ -0,0 +1,222 @@
+<?php
+
+/**
+ * Generic Netscape bookmark parser
+ */
+class NetscapeBookmarkParser
+{
+    protected $defaultTag = '';
+    protected $defaultPub = false;
+    protected $items = array();
+
+    const TRUE_PATTERN = 'y|yes|on|checked|ok|1|true|array|\+|okay|yes|t|one';
+    const FALSE_PATTERN = 'n|no|off|empty|null|false|nil|0|-|exit|die|neg|f|zero|void';
+
+    /**
+     * Instantiates a new NetscapeBookmarkParser
+     *
+     * @param string $defaultTag Link tag if missing
+     * @param mixed  $defaultPub Link publication status if missing
+     *                           - 'true' => public
+     *                           - 'false' => private)
+     */
+    public function __construct($defaultTag=null, $defaultPub=false)
+    {
+        if ($defaultTag) {
+            $this->defaultTag = $defaultTag;
+        } else {
+            $this->defaultTag = 'imported-'.date("Ymd");
+        }
+        if ($defaultPub) {
+            $this->defaultPub = $defaultPub;
+        }
+    }
+
+    /**
+     * Parses a Netscape bookmark file
+     *
+     * @param string $filename Bookmark file to parse
+     *
+     * @return array An associative array containing parsed links
+     */
+    public function parseFile($filename)
+    {
+        return $this->parseString(file_get_contents($filename));
+    }
+
+    /**
+     * Parses a string containing Netscape-formatted bookmarks
+     *
+     * Output format:
+     *
+     *     Array
+     *     (
+     *         [0] => Array
+     *             (
+     *                 [note]  => Some comments about this link
+     *                 [pub]   => 1
+     *                 [tags]  => a list of tags
+     *                 [time]  => 1459371397
+     *                 [title] => Some page
+     *                 [uri]   => http://domain.tld:5678/some-page.html
+     *             )
+     *         [1] => Array
+     *             (
+     *                 ...
+     *             )
+     *     )
+     *
+     * @param string $bookmarkString String containing Netscape bookmarks
+     *
+     * @return array An associative array containing parsed links
+     */
+    public function parseString($bookmarkString) {
+        $i = 0;
+        $next = false;
+        $currentTag = $this->defaultTag;
+
+        $lines = explode("\n", $this->sanitizeString($bookmarkString));
+
+        foreach ($lines as $line_no => $line) {
+            if (preg_match('/^<h\d(.*?)>(.*?)<\/h\d>/i', $line, $m1)) {
+                // a tag is matched: set its value as current
+                $currentTag = $m1[2];
```

Comment:
Looking at `https://github.com/shaarli/netscape-bookmark-parser/blob/shaarli-devel/tests/input/shaarli.htm#L7`, I'm not sure that's a good idea.